libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp

```diff
@@ -236,75 +368,145 @@ namespace hpx { namespace parallel { inline namespace v1 {
-        FwdIter1 last, FwdIter2 dest_first)
+        rotate_copy(ExPolicy&& policy, FwdIter1 first, FwdIter1 new_first,
+            FwdIter1 last, FwdIter2 dest_first)
     {
-        static_assert((hpx::traits::is_forward_iterator<FwdIter1>::value),
+        static_assert((hpx::traits::is_forward_iterator_v<FwdIter1>),
             "Requires at least forward iterator.");
-        static_assert((hpx::traits::is_forward_iterator<FwdIter2>::value),
+        static_assert((hpx::traits::is_forward_iterator_v<FwdIter2>),
             "Requires at least forward iterator.");
 
-        typedef std::integral_constant<bool,
+        using is_seq = std::integral_constant<bool,
             hpx::is_sequenced_execution_policy<ExPolicy>::value ||
-                !hpx::traits::is_bidirectional_iterator<FwdIter1>::value>
-            is_seq;
+                !hpx::traits::is_bidirectional_iterator_v<FwdIter1>>;
 
         return detail::rotate_copy<util::in_out_result<FwdIter1, FwdIter2>>()
             .call2(std::forward<ExPolicy>(policy), is_seq(), first, new_first,
                 last, dest_first);
     }
 }}}    // namespace hpx::parallel::v1
+
+// create new APIs, tag_fallback_dispatch overloads.
+namespace hpx {
+    ///////////////////////////////////////////////////////////////////////////
+    // CPO for hpx::rotate
+    HPX_INLINE_CONSTEXPR_VARIABLE struct rotate_t final
+      : hpx::functional::tag_fallback<rotate_t>
+    {
+        // clang-format off
+        template <typename FwdIter,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator_v<FwdIter>
+            )>
+        // clang-format on
+        friend FwdIter tag_fallback_dispatch(
+            hpx::rotate_t, FwdIter first, FwdIter new_first, FwdIter last)
+        {
+            static_assert(hpx::traits::is_forward_iterator_v<FwdIter>,
+                "Requires at least forward iterator.");
+
+            return parallel::util::get_second_element(
+                hpx::parallel::v1::detail::rotate<
+                    hpx::parallel::util::in_out_result<FwdIter, FwdIter>>()
+                    .call(hpx::execution::seq, first, new_first, last));
+        }
+        // clang-format off
+        template <typename ExPolicy, typename FwdIter,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::is_execution_policy<ExPolicy>::value &&
+                hpx::traits::is_iterator_v<FwdIter>
+            )>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy,
+            FwdIter>::type
+        tag_fallback_dispatch(hpx::rotate_t, ExPolicy&& policy, FwdIter first,
+            FwdIter new_first, FwdIter last)
+        {
+            static_assert((hpx::traits::is_forward_iterator_v<FwdIter>),
+                "Requires at least forward iterator.");
+
+            using is_seq = std::integral_constant<bool,
+                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
```

Comment:
FWIW, we have a `hpx::is_sequenced_execution_policy_v<ExPolicy>` as well (https://github.com/STEllAR-GROUP/hpx/blob/master/libs/parallelism/execution/include/hpx/execution/traits/is_execution_policy.hpp#L116-L118).

Reply:
Got it, will check and replace it.

Reply:
Got it, will check and replace it.